File: dotnet/DesignPatterns/StructuralDesignPatterns/DecoratorDesignPattern/DecoratorDesignPattern.Example1/ArmorDecorator.cs
namespace DecoratorDesignPattern.Example1;

public class ArmorDecorator : CharacterDecorator
{
    private readonly string _armorName;
    private readonly int _defenseBonus;
    private readonly int _healthBonus;

    public ArmorDecorator(ICharacter character, string armorName, int defenseBonus, int healthBonus)
        : base(character)
    {
        if (defenseBonus <= 0 || healthBonus <= 0)
            throw new ArgumentException("Bonuslar pozitif olmalıdır.");

        _armorName = armorName ?? "Koruyucu Zırh";
        _defenseBonus = defenseBonus;
        _healthBonus = healthBonus;
    }

    public override int GetDefense()
    {
        return base.GetDefense() + _defenseBonus;
    }

    public override int GetHealth()
    {
        return base.GetHealth() + _healthBonus;
    }

    public override string GetDescription()
    {
        return $"{base.GetDescription()} + {_armorName} (+{_defenseBonus} Savunma, +{_healthBonus} Can)";
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/DecoratorDesignPattern/DecoratorDesignPattern.Example1/src/ArmorDecorator.ts
import { ICharacter } from './ICharacter';
import { CharacterDecorator } from './CharacterDecorator';

export class ArmorDecorator extends CharacterDecorator {
  constructor(
    character: ICharacter,
    private readonly armorName: string,
    private readonly defenseBonus: number,
    private readonly healthBonus: number,
  ) {
    super(character);
    if (defenseBonus <= 0 || healthBonus <= 0) throw new Error('Bonuslar pozitif olmalıdır.');
  }

  getDefense(): number { return super.getDefense() + this.defenseBonus; }
  getHealth(): number { return super.getHealth() + this.healthBonus; }
  getDescription(): string {
    return `${super.getDescription()} + ${this.armorName} (+${this.defenseBonus} Savunma, +${this.healthBonus} Can)`;
  }
}
